modules/table/src/main/java/org/apache/ignite/internal/table/distributed/gc/MvGc.java

```diff
@@ -221,34 +222,62 @@ private void scheduleGcForStorage(TablePartitionId tablePartitionId) {
                 return;
             }
 
-            if (storageHandler.gcInProgressFuture.get() != future) {
+            if (storageHandler.gcInProgressFuture.get() != currentGcFuture) {
                 // Someone in parallel is already collecting garbage, we will try once again after completion of gcInProgressFuture.
                 return;
             }
 
             try {
-                for (int i = 0; i < GC_BATCH_SIZE; i++) {
-                    HybridTimestamp lowWatermark = lowWatermarkReference.get();
+                HybridTimestamp lowWatermark = lowWatermarkReference.get();
 
-                    assert lowWatermark != null : tablePartitionId;
+                assert lowWatermark != null : tablePartitionId;
 
-                    // If storage has been deleted or there is no garbage, then for now we will stop collecting garbage for this storage.
-                    if (!storageHandlerByPartitionId.containsKey(tablePartitionId)
-                            || !storageHandler.storageUpdateHandler.vacuum(lowWatermark)) {
-                        return;
-                    }
-                }
-            } catch (Throwable t) {
-                future.completeExceptionally(t);
+                // If the storage has been deleted, then garbage collection is no longer necessary.
+                if (!storageHandlerByPartitionId.containsKey(tablePartitionId)) {
+                    currentGcFuture.complete(null);
 
-                return;
-            } finally {
-                if (!future.isCompletedExceptionally()) {
-                    future.complete(null);
+                    return;
                 }
-            }
 
-            scheduleGcForStorage(tablePartitionId);
+                StorageUpdateHandler storageUpdateHandler = storageHandler.storageUpdateHandler;
+
+                // We can only start garbage collection when the partition safe time is reached.
+                storageUpdateHandler.getSafeTimeTracker()
+                        .waitFor(lowWatermark)
+                        .thenApplyAsync(unused -> {
+                            for (int i = 0; i < GC_BATCH_SIZE; i++) {
+                                // If the storage has been deleted or there is no garbage, then we will stop.
+                                if (!storageHandlerByPartitionId.containsKey(tablePartitionId)
+                                        || !storageUpdateHandler.vacuum(lowWatermark)) {
+                                    return false;
+                                }
+                            }
+
+                            return true;
+                        }, executor)
+                        .whenComplete((isLeftGarbage, throwable) -> {
+                            if (throwable != null) {
+                                if (throwable instanceof TrackerClosedException
```

Comment:
I believe only one of these options is actually possible. Why do you have both?

Reply:
If you run code something like this, you'll see an error like this:
```
        CompletableFuture<?> future = new CompletableFuture<>();

        future.thenAcceptAsync(unused -> {}).whenComplete((unused, throwable) -> throwable.printStackTrace());

        future.completeExceptionally(new TrackerClosedException());
```

`java.util.concurrent.CompletionException: org.apache.ignite.internal.util.TrackerClosedException`